pandas/core/groupby/generic.py

```diff
@@ -355,18 +270,48 @@ def _aggregate_item_by_item(self, func, *args, **kwargs):
 
         return DataFrame(result, columns=result_columns)
 
-    def _decide_output_index(self, output, labels):
-        if len(output) == len(labels):
-            output_keys = labels
+    def _decide_output_index(
+        self,
+        output: Dict,
+        labels: Index,
+        col_labels: Optional[List[Union[Hashable, Tuple[Hashable, ...]]]] = None,
+    ) -> Index:
+        """
+        Determine axis labels to use while wrapping aggregated values.
+
+        Parameters
+        ----------
+        output : dict of ndarrays
+            Results of aggregating by-column.
+        labels : Index
+            Existing labels of selected object. Used to determine resulting
+            shape and name(s).
+        col_labels : list, optional
+            The ultimate column labels for the reshaped object. Each entry in
+            this list should correspond to a key value in output. Must be valid
+            column labels and tuples are contained within should map to a
+            MultiIndex.
+
+        Returns
+        -------
+        Index or MultiIndex
+
+        Notes
+        -----
+        Ideally output should always have integers as a key and the col_labels
```

Comment:
Added this as a note for now because I didn't want duplicate labels to complicate the diff, but this should be bundled with a more comprehensive handling of iteration that I think will need to go column-by-column

Here's an example where we iterate by "slice":

https://github.com/pandas-dev/pandas/blob/558f754e1787625d8b18cd4960fa059d3533faae/pandas/core/groupby/groupby.py#L2263

But a slice can still be a DataFrame in case of duplicated column labels, and especially if those duplicate labels don't have a contiguous dtype can make things messy. 

I think comprehensively need to change the iterations in these modules and ensure consistency